text/temp/PALENICA/Larsson_Sadakane_Suffix_Array.tex

```diff
@@ -0,0 +1,105 @@
+%! Author = pawel
+%! Date = 6/17/20
+
+% Preamble
+\documentclass[12pt]{article}
+
+% Packages
+\usepackage[T1]{fontenc}
+\usepackage[utf8]{inputenc}
+\usepackage{amsthm}
+\usepackage{amsmath} 
+\usepackage{hyperref}
+\usepackage{cleveref}
+\usepackage{sectsty}
+\usepackage{algorithmicx,algpseudocode}
+\usepackage{titling}
+\usepackage[
+backend=biber,
+style=alphabetic,
+sorting=ynt
+]{biblatex}
+\addbibresource{larsson_sadakane.bib}
+
+\newtheorem*{observation}{Obserwacja}
+\newtheorem*{definition}{Definicja}
+\newtheorem*{lemma}{Lemat}
+
+% Commands
+\newcommand{\subtitle}[1]{%
+  \posttitle{%
+    \par\end{center}
+    \begin{center}\large#1\end{center}
+    \vskip0.5em}%
+}
+
+% Document
+\begin{document}
+\title{Faster suffix sorting}
+\subtitle{Algorytm obliczania tablicy sufiksowej autorstwa N. Larssona i K. Sadakane \cite{fastsufsort}}
+\author{Paweł Palenica}
+\date{}
+\maketitle
+
+\section*{Wstęp}
+Algorytmy konstrukcji tablicy sufiksowej są jednym z klasycznych zagadnień algorytmów tekstowych. Podczas gdy znane są algorytmy tworzące tę strukturę w czasie liniowym od długości słowa wejściowego, często wymagają one dużego narzutu pamięciowego. Zastosowaniem tablicy sufiksowej dla którego duży narzut pamięciowy wynikający z użycia np. drzewa sufiksowego może być niepożądany jest transformata Burrowsa-Wheelera - odwracalna transformacja tekstu która zmienia kolejność znaków w tekście aby zwiększyć ciągi podobnych znaków, co może pomóc w kompresji.
+
+Problem który będzie rozwiązywany definiujemy następująco. Niech $X = x_1 x_2 \ldots x_n \$ $ będzie ciągiem znaków złożonym z ciągu wejściowego długości $n$ z dopisanym symbolem $\$$, który jest leksykograficznie przed wszystkimi symbolami ciągu wejściowego. Jako $S_i,\ 1 \leq i \leq n+1$ będziemy oznaczać sufiks $X$ zaczynający się w pozycji $i$. Wyjściem algorytmu ma być tablica indeksów $I$ (\textit{tablica sufiksowa}) taka że $S_{I[k-1]}$ jest leksykograficznie mniejsze od $S_{I[k]}$ dla $0 < k \leq n$.
+
+Algorytm tworzenia tablicy sufiksowej zaprezentowany przez N. Larssona i K. Sadakane działa w czasie $O(n\ log\ n)$ gdzie $n$ jest długością wejśćiowego tekstu. Autorzy rozbudowują podejście \cite{basicAlgo} i korzystają z techniki `podwajania` znanej z algorytmu Karpa, Millera, Rosenbera która polega na użyciu pozycji sufiksów po fazie sortowania jako kluczy do sortowania dwukrotnie dłuższych sufiksów w kolejnej fazie. Autorzy formalizują to podejście definiując pojęcie \textit{h-order}, czyli porządek leksykograficzny sufiksów rozważając jedynie $h$ początkowych znaków każdego sufiksu. Zayważmy, że \textit{h-order} niekoniecznie jest jednoznaczny jeśli $h < n$.
```

Comment:
Well I changed the language of spellcheck in overleaf only after ~half of this doc was written (that's why no spelling errors in later parts) and proofread only the pdf instead of raw tex. Good catch, fixed.